dev/PushNotifications/PushNotificationsLongRunningTask/ForegroundSinkManager.h

```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include <unordered_map>
+#include <unordered_set>
+#include <algorithm>
+#include <vector>
+
+class ForegroundSinkManager
+{
+public:
+    ForegroundSinkManager() = default;
+
+    void Add(std::wstring const& processName, IWpnForegroundSink* const& sink);
+
+    void Remove(std::wstring const& processName);
+
+    bool InvokeForegroundHandlers(std::wstring const& processName, winrt::com_array<uint8_t> const& payload, ULONG const& payloadSize);
+
+private:
+    // An app can only have one activate foreground sink with Long Running Process. Event handlers in
+    // the sink are managed by the WindowsAppSDK.
+    std::unordered_map<std::wstring, Microsoft::WRL::ComPtr<IWpnForegroundSink>> m_foregroundMap = {};
```

Comment:
@pmpurifoy There is a bug here. The map needs to technically hold an AgileRef of IWpnForegroundSink. Otherwise, we will have issues invoking the callback if apartment contexts don't match in client/server.

Reply:
PushNotificationsChannel is a winrt object and by default agile. Do I need to cast it to IAgileReference then send it to the LRP?

Reply:
If it is already Agile, then I don't believe we need to do anything here. + @jonwis for clarification for best practises.

Reply:
Here is a link to winrt default agile behavior.
https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/agile-objects

Reply:
yeah. I think you are good then. Maybe do the IAgileObject conversion and store that in the map: winrt::com_ptr<IAgileObject> iagileobject{ myimpl.as<IAgileObject>() };

But if winrt gives this to you for free, this may not even be necessary since your implementation is taken care of.